plugins/bufferlines/barbar.nix

```diff
@@ -69,249 +24,566 @@ let
-    enable = mkEnableOption "barbar.nvim";
+helpers.neovim-plugin.mkNeovimPlugin config {
+  name = "barbar";
+  originalName = "barbar.nvim";
+  defaultPackage = pkgs.vimPlugins.barbar-nvim;
+  extraPlugins = [ pkgs.vimPlugins.nvim-web-devicons ];
+
+  maintainers = [ maintainers.GaetanLepage ];
+
+  # TODO: introduced 2024-05-30, remove on 2024-07-30
+  deprecateExtraOptions = true;
+  optionsRenamedToSettings = [
+    "animation"
+    "autoHide"
+    "clickable"
+    "focusOnClose"
+    [
+      "hide"
+      "alternate"
+    ]
+    [
+      "hide"
+      "current"
+    ]
+    [
+      "hide"
+      "extensions"
+    ]
+    [
+      "hide"
+      "inactive"
+    ]
+    [
+      "hide"
+      "visible"
+    ]
+    "highlightAlternate"
+    "highlightInactiveFileIcons"
+    "highlightVisible"
+    [
+      "icons"
+      "bufferIndex"
+    ]
+    [
+      "icons"
+      "bufferNumber"
+    ]
+    [
+      "icons"
+      "button"
+    ]
+    [
+      "icons"
+      "filetype"
+      "customColors"
+    ]
+    [
+      "icons"
+      "separator"
+      "left"
+    ]
+    [
+      "icons"
+      "separator"
+      "right"
+    ]
+    "insertAtStart"
+    "insertAtEnd"
+    "maximumPadding"
+    "minimumPadding"
+    "maximumLength"
+    "semanticLetters"
+    "letters"
+    "sidebarFiletypes"
+    "noNameTitle"
+    "tabpages"
+  ];
+  imports =
+    let
+      basePluginPath = [
+        "plugins"
+        "barbar"
+      ];
+      settingsPath = basePluginPath ++ [ "settings" ];
+    in
+    [
+      (mkRenamedOptionModule (basePluginPath ++ [ "excludeFileTypes" ]) (
+        settingsPath ++ [ "exclude_ft" ]
+      ))
+      (mkRenamedOptionModule (basePluginPath ++ [ "excludeFileNames" ]) (
+        settingsPath ++ [ "exclude_name" ]
+      ))
+      (mkRemovedOptionModule (
+        basePluginPath
+        ++ [
+          "icons"
+          "diagnostics"
+        ]
+      ) "Use `settings.icons.diagnostics` instead, but pay attention as the keys have changed.")
+      (mkRenamedOptionModule
+        (
+          basePluginPath
+          ++ [
+            "icons"
+            "filetype"
+            "enable"
+          ]
+        )
+        (
+          settingsPath
+          ++ [
+            "icons"
+            "filetype"
+            "enabled"
+          ]
+        )
+      )
+    ]
+    ++ (map
+      (
+        name:
+        mkRemovedOptionModule (
+          basePluginPath
+          ++ [
+            "icons"
+            name
+          ]
+        ) "Use `settings.icons.${name}` instead, but you should now use the real `snake_case` key names."
+      )
+      [
+        "alternate"
+        "current"
+        "inactive"
+        "modified"
+        "pinned"
+        "visible"
+      ]
+    );
+
+  extraOptions = {
+    keymaps =
+      {
+        silent = mkEnableOption "silent keymaps for barbar";
+      }
+      // (mapAttrs (
+        optionName: funcName: helpers.mkNullOrOption types.str "Keymap for function Buffer${funcName}"
```

Comment:
Could this be type `(either str keymap-submodule)`? Maybe the user wants a different `mode` or a custom `options.desc`, etc

Reply:
`either str` or this:

```nix
(submodule {
  options = {
    mode = keymap-helpers.mkModeOption "n";
    options = keymap-helpers.mapConfigOptions;
  };
})
```

?

Reply:
Are we sure that the plugin accepts this ?

Reply:
> Are we sure that the plugin accepts this ?

This is stupid, it is not a plugin option.

Reply:
We could also do this first: https://github.com/nix-community/nixvim/issues/603#issuecomment-2143472314

Reply:
As this PR is quite big, I would have such a change in a dedicated one.

Reply:
https://github.com/nix-community/nixvim/pull/1616

Reply:
Done.